ssh.go

```diff
@@ -6,24 +6,113 @@ import (
 	"fmt"
 	"io"
 	"log"
+	"net"
 	"os"
 	"os/exec"
+	"strconv"
 	"strings"
 	"sync"
 	"syscall"
 	"unsafe"
 
+	"github.com/gokrazy/gokrazy"
 	"github.com/google/shlex"
 	"github.com/kr/pty"
 	"golang.org/x/crypto/ssh"
 )
 
-func handleChannel(newChannel ssh.NewChannel) {
-	if t := newChannel.ChannelType(); t != "session" {
-		newChannel.Reject(ssh.UnknownChannelType, fmt.Sprintf("unknown channel type: %q", t))
+func handleChannel(newChan ssh.NewChannel) {
+	switch t := newChan.ChannelType(); t {
+	case "session":
+		handleSession(newChan)
+	case "direct-tcpip":
+		handleTCPIP(newChan)
+	default:
+		newChan.Reject(ssh.UnknownChannelType, fmt.Sprintf("unknown channel type: %q", t))
+		return
+	}
+}
+
+func parseAddr(addr string) net.IP {
+	ip := net.ParseIP(addr)
+	if ip == nil {
+		if ips, err := net.LookupIP(addr); err == nil {
+			ip = ips[0] // use first address found
+		}
+	}
+
+	return ip
+}
+
+// Forwarding ported from https://github.com/gliderlabs/ssh (BSD3 License)
+
+// direct-tcpip data struct as specified in RFC4254, Section 7.2
+type localForwardChannelData struct {
+	DestAddr string
+	DestPort uint32
+
+	OriginAddr string
+	OriginPort uint32
+}
+
+func handleTCPIP(newChan ssh.NewChannel) {
+	d := localForwardChannelData{}
+	if err := ssh.Unmarshal(newChan.ExtraData(), &d); err != nil {
+		newChan.Reject(ssh.ConnectionFailed, "error parsing forward data: "+err.Error())
+		return
+	}
+
+	var ip net.IP
+	switch *forwarding {
+	case "loopback":
+		if ip := parseAddr(d.DestAddr); ip != nil && !ip.IsLoopback() {
```

Comment:
Here and below: you’re using `ip := `, so you’re creating a new `ip` variable and the one declared above will remain nil, meaning the rest of the code will not work. Have you not tested this PR?

Reply:
> Have you not tested this PR?

Not after the latest changes :(

Reply:
Problem ist that with the dependencies on internal I cannot test this in my own package as it won't compile. It's a PITA :(

Reply:
> Problem ist that with the dependencies on internal I cannot test this in my own package as it won't compile, at least not easily. It's a PITA :(

Can you not check out your fork under the original name github.com/gokrazy/breakglass to make the internal repositories work?

Reply:
How would I make `gokr-packer` find that for uploading? I need to test this on-device...

Reply:
> How would I make `gokr-packer` find that for uploading? I need to test this on-device...

In your go.mod, add a replace statement: 

* https://golang.org/ref/mod#go-mod-file-replace
* https://github.com/golang/go/wiki/Modules#when-should-i-use-the-replace-directive